dist_run.py

```diff
@@ -6,67 +6,151 @@
+
+    hf_config = get_hf_config_file(hf_model_name)
+    if hf_config is None:
+        raise ValueError(f"Config file not found for model id {hf_model_name}")
+    logger.info(f"Using HF model weights from {hf_model_name}")
+
+    mesh_dimensions = (2, 2)
+    device_mesh = _create_device_mesh(mesh_dimensions)
 
-    # Construct a device mesh with available devices (multi-host or single host)
-    device_mesh = dist.init_device_mesh("cuda", (2, 2), mesh_dim_names=("pp", "tp"))
     tp_mesh = device_mesh["tp"]
     pp_mesh = device_mesh["pp"]
     pp_rank = pp_mesh.get_local_rank()
     nstages = pp_mesh.size()
-
-    rank = dist.get_rank()
     device = torch.device(f"cuda:{rank}")
 
-    # Create parallel model with device_mesh context
     with device:
         with tp_mesh:
             model = TransformerStage(config, pp_rank, nstages)
             model.setup_caches(1, 4096)
+            # TODO: refine this .to once we start using fp8 for KV cache
+            model = model.to(model_dtype)
+    logger.info(f"Model: {model}")
 
-    print(model)
+    mbs = 2  # number of micro-batches
+    mb_size = 1  # micro-batch size
+    batch_size = mbs * mb_size  # total batch size
+    seqlen = 4096  # sequence length
+    dim = 4096  # embedding dimension
 
-    # Distributed run
-    mbs = 2                         # number of micro-batches
-    mb_size = 1                     # micro-batch size
-    batch_size = mbs * mb_size      # total batch size
-    seqlen = 4096                   # sequence length
-    dim = 4096                      # embedding dimension
-
-    # Example input for pipeline stages
     mb_ids = torch.randint(0, config.vocab_size, (mb_size, seqlen), device=device)
-    activation = torch.rand(mb_size, seqlen, dim, device=device)
+    activation = torch.rand(mb_size, seqlen, dim, device=device, dtype=model_dtype)
     example_args = mb_ids if pp_rank == 0 else activation
 
-    # Create pipeline stages
+    logger.info(f"Creating pipeline stage {pp_rank=}, {nstages=}")
     stage = PipelineStage(
-        model, pp_rank, nstages, device,
+        model,
+        pp_rank,
+        nstages,
+        device,
         input_args=(example_args,),
         group=pp_mesh.get_group(),
     )
 
-    # Run pipeline
-    schedule = ScheduleGPipe(stage, mbs)
+    cpu_tensors = find_cpu_tensors(stage.submod)
+    if len(cpu_tensors) > 0:
+        raise ValueError("Found cpu tensors in stage")
+
+    # Load weights
+    logger.info(f"Loading weights for {pp_rank=} on {device=}")
+    _load_model_weights(stage.submod, hf_model_name, device=device, logger=logger)
+
+    stage.submod.eval()
+
+    # this check confirms that there are no cpu tensors in the model..we expect this to be true.
+    cpu_tensors = find_cpu_tensors(stage.submod)
+    # logger.info(f"Found {len(cpu_tensors)} cpu tensors: {cpu_tensors}")
+    if len(cpu_tensors) > 0:
+        raise ValueError("Found cpu tensors in stage")
```

Comment:
duplicate code?

Reply:
I had that there intentionally earlier (to check pre and post) but you are right, we don't need it x2 now that things are working. 
Have removed to just leave the second one as a verification after loading.